csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,584 @@
+            // Timestamp type
+            public void Visit(TimestampType datatype)
+            {
+                FormatString = String.Format("ts{0}:{1}", TimeUnitComponent(datatype.Unit), datatype.Timezone);
+            }
+            // Nested
+            public void Visit(ListType _) => FormatString = "+l";
+            public void Visit(StructType _) => FormatString = "+s";
+            // Dictionary
+            public void Visit(DictionaryType datatype)
+            {
+                // format string is that of the indices
+                datatype.IndexType.Accept(this);
+            }
+
+            // Unsupported
+            public void Visit(IArrowType type)
+            {
+                throw new NotImplementedException($"Exporting {type.Name} not implemented");
+            }
+        }
+
+        private sealed class ImportedArrowSchema : IDisposable
+        {
+            private readonly CArrowSchema _data;
+            private readonly IntPtr _handle;
+            private readonly bool _isRoot;
+
+            public ImportedArrowSchema(IntPtr handle)
+            {
+                _data = Marshal.PtrToStructure<CArrowSchema>(handle);
+                if (_data.release == null)
+                {
+                    throw new Exception("Tried to import a schema that has already been released.");
+                }
+                _handle = handle;
+                _isRoot = true;
+            }
+
+            private ImportedArrowSchema(IntPtr handle, bool isRoot) : this(handle)
+            {
+                _isRoot = isRoot;
+            }
+
+            public void Dispose()
+            {
+                // We only call release on a root-level schema, not child ones.
+                if (_isRoot)
+                {
+                    _data.release(_handle);
+                }
+            }
+
+            public ArrowType GetAsType()
+            {
+                var format = StringUtil.PtrToStringUtf8(_data.format);
+                if (_data.dictionary != IntPtr.Zero)
+                {
+                    ArrowType indicesType = format switch
+                    {
+                        "c" => new Int8Type(),
+                        "C" => new UInt8Type(),
+                        "s" => new Int16Type(),
+                        "S" => new UInt16Type(),
+                        "i" => new Int32Type(),
+                        "I" => new UInt32Type(),
+                        "l" => new Int64Type(),
+                        "L" => new UInt64Type(),
+                        _ => throw new InvalidDataException($"Indices must be an integer, but got format string {format}"),
+                    };
+
+                    var dictionarySchema = new ImportedArrowSchema(_data.dictionary, /*is_root*/ false);
+                    ArrowType dictionaryType = dictionarySchema.GetAsType();
+
+                    bool ordered = (_data.flags & CArrowSchema.ArrowFlagNullable) == CArrowSchema.ArrowFlagNullable;
```

Comment:
ArrowFlagDictionaryOrdered?

Reply:
Thanks. I've updated the tests to catch this.